### ISSUE TITLE
[bug]: Rectangle "mousedown" persists when cursor moved beyond canvas

### ISSUE THREAD

**Issue by Fortyseven** [invoke-ai/InvokeAI#6306]
### Is there an existing issue for this problem?

- [X] I have searched the existing issues

### Operating system

Linux

### GPU vendor

Nvidia (CUDA)

### GPU model

4090

### GPU VRAM

24

### Version number

4.20b1

### Browser

Firefox 125.0.3 

### What happened

Rectangle stays engaged with the cursor, undrawn, as if it never receives a 'mouseup' event. Need to click inside the canvas, separately, to cancel it.

### What you expected to happen

Rectangle should stop at the bounds of the canvas (preferably leaving the rectangle rendered as opposed to cancelling it entirely, so it's easy to fill to the edges).

### How to reproduce the problem

Go to Control Layer canvas, start drawing rectangle, click and drag off the canvas.

### Additional context

![image](https://github.com/invoke-ai/InvokeAI/assets/156409/bda4ffae-2026-4cd9-b9f0-5024aa67c9c8)

### Discord username

fortyseven

**Comment by psychedelicious** [comment]
I agree it should work like this but it's rather finicky. Mouse events are handled by the canvas itself and moving the mouse _outside_ the canvas doesn't trigger them, so we don't get `mouseup` or `mousemove` events when the cursor leaves.  We do get a `mouseleave` with some coordinates.

The way to handle this is to snap to the edges of the canvas when using the rectangle tool, and not hide the rect if the mouse leaves the stage. This was a bit much to implement at the early stage when I added this tool but now that the dust has settled it shouldn't be too hard to do.

**Comment by psychedelicious** [comment]
This is partially resolved by #6311. The rect snaps to the edges of the canvas.
